app/grandchallenge/algorithms/forms.py

```diff
@@ -1369,11 +1396,31 @@ def clean(self):
                 f"{oxford_comma(duplicate_interfaces)} present in both"
             )
 
-        cleaned_data["existing_io"] = (
-            get_existing_interface_for_inputs_and_outputs(
-                inputs=inputs,
-                outputs=outputs,
-            )
+        return cleaned_data
+
+    def save(self):
+        interface = AlgorithmInterface.objects.create(
+            inputs=self.cleaned_data["inputs"],
+            outputs=self.cleaned_data["outputs"],
         )
 
-        return cleaned_data
+        if self.cleaned_data["set_as_default"]:
+            AlgorithmAlgorithmInterface.objects.filter(
+                algorithm=self._algorithm
+            ).update(is_default=False)
+
+        if AlgorithmAlgorithmInterface.objects.filter(
+            algorithm=self._algorithm, interface=interface
+        ).exists():
+            AlgorithmAlgorithmInterface.objects.filter(
+                algorithm=self._algorithm, interface=interface
+            ).update(is_default=self.cleaned_data["set_as_default"])
+        else:
+            self._algorithm.interfaces.add(
+                interface,
+                through_defaults={
+                    "is_default": self.cleaned_data["set_as_default"]
+                },
+            )
```

Comment:
`update` returns the number of changed objects, so I think you can do this in one call (swapping a LBYL for an EAFP):

```suggestion
        affected_rows = AlgorithmAlgorithmInterface.objects.filter(
                algorithm=self._algorithm, interface=interface
            ).update(is_default=self.cleaned_data["set_as_default"])
        
        if affected_rows == 0:
            self._algorithm.interfaces.add(
                interface,
                through_defaults={
                    "is_default": self.cleaned_data["set_as_default"]
                },
            )
        elif affected_rows > 1:
            raise RuntimeError("This _really_ shouldn't happen")
```

Reply:
Actually - not sure about that - does `affected_rows` return the number of objects changed or the number of objects in the filter?

Reply:
I asked ChatGPT and it suggests that it does behave in the same way that I originally suspected so this should work - the `affected_rows` is the number of rows in the filter regardless of whether `is_default` was changed. But, good to double check this.

Reply:
As per the documentation: 

> The update() method is applied instantly and returns the number of rows matched by the query (which may not be equal to the number of rows updated if some rows already have the new value).

But that doesn't matter. Your code still works since there should always be either 1 or 0 matching rows.

Reply:
It just semantically is no longer correct to call them `affected_rows`, maybe `rows_to_update` is better?

Reply:
I don't think `rows_to_update` works as the tense is misleading - the updating has already happened by the time the value is assigned. I suggested affected rows as that is what the Django docs calls them, but maybe `updated_rows` or `number_of_updated_interfaces` is clearer.

Reply:
But that suggests that it reflects the number of rows that were actually updated, which is not what update returns. Going by what the Django documentation calls them, `matched_rows` is closest to describing what `update` returns.